### DISCUSSION TITLE
Parsing multiple-language documents

### DISCUSSION THREAD

**Discussion by vigoux** [tree-sitter/tree-sitter#793] [Q&A] [answered]
Hi, I am one of the maintainers of `nvim-treesitter` an org to enhance syntax highlighting in Neovim, using Tree-sitter.

Lately we (together with @steelsojka) have been working on multiple-language documents, and correctly parsing those.
For most cases we successfully did it, but one weird case still remain.

Let's say I have a Rust file with doc comments in it. Those comments needs to be parsed using the markdown parser, but they need to be parse independently.
The way we are currently doing it is by setting the included ranges of a parser to only on of the comments, parse the comment, get the tree, and repeat for another comment.

We can't do in any other way, because if we set the included ranges of the parser to be the whole set of ranges, it will parse them as one.

In my opinion this is some kind of hack, so what is the recommended way to go around this problem ?

Thanks in advance.

**Comment by maxbrunsfeld** [comment, marked as answer]
Hi @vigoux, this is a good question. I'm not 100% sure what is the best approach for parsing doc comments. But I do have a few thoughts:

> The way we are currently doing it is by setting the included ranges of a parser to only on of the comments, parse the comment, get the tree, and repeat for another comment.

When using Tree-sitter for syntax highlighting in Atom (which was implemented quite a long time ago now, and doesn't use tree queries), we did something similar to this. For JavaScript, every individual comment was parsed as its own JSDoc tree. It isn't semantically *correct*, because really, a group of adjacent doc comments constitute one single "document". But it was good enough for syntax highlighting purposes, and that was all we needed.

> We can't do in any other way, because if we set the included ranges of the parser to be the whole set of ranges, it will parse them as one.

I think you can do this. You would need to:
  * Identify adjacent *sequences* of comments (e.g. all of the adjacent doc comments before a function)

    ```clj
    ((comment)+ @doc . (function_item))
    ```

  * For each match, select only the final span of comments with no blank lines in between them ☝️ 
  * Get all of these comments' ranges
  * Parse all of those ranges together as one document, using `ts_parser_set_included_ranges`.

This should work. You won't be able to test your code with the `tree-sitter highlight` command, because currently the `tree-sitter-highlight` crate does not have a way to specify language injection within arbitrary *groups* of nodes like this. It actually did before, but we then [simplified it](https://github.com/tree-sitter/tree-sitter/pull/526) in order to handle PHP. But we could probably re-generalize it with a slightly different API so that you could do this.

> what is the recommended way to go around this problem ?

On the other hand, this approach may be more complex than is needed. In the case of doc comments, we might only need very simple highlighting, and not other syntax-based features. So it may be good enough to just parse each one individually.

Also, because these syntax trees are so small, it may not be worth maintaining them as state in between frames. It may be simpler and faster to just re-parse them from scratch on demand when rendering highlights. Does that make sense?

**Reply by vigoux** [reply]
I think it makes sense

**Reply by vigoux** [reply]
So, just to sum up, the solution is the retain some trees, on for each comment or whatever it is, and keeps them in sync, right ?

**Reply by maxbrunsfeld** [reply]
I don't know; I would sum up my thoughts as: there are a few different valid approaches, I'm not sure what tradeoffs make sense in the context of Neovim.

I'm very glad to see that you're exploring this space though. Thanks for sharing your thoughts your status with it!

**Reply by vigoux** [reply]
Thanks for your help ! We are actually going through a different approach than Atom's language layers, by using a language tree for the document, we'll see how it goes !

**Reply by steelsojka** [reply]
@maxbrunsfeld If I could just add to this. I currently run into a problem parsing all the comments as a single document. Here is an example.
![Screen Shot 2020-11-03 at 2 06 10 PM](https://user-images.githubusercontent.com/2062154/98113018-a8103a80-1e68-11eb-9757-44b0939b338a.png)

This is using the jsdoc parser and parsing both the comments as one document. This was the main motivation to doing them as separate documents. Is this just an issue with the JSdoc parser?

**Reply by maxbrunsfeld** [reply]
Yeah, the JSDoc parser just doesn’t allow multiple multi-line comments in a row like that. I think it works with multiple single line (‘//‘) comments though. We might need to generalize it to allow code like that.

**Reply by steelsojka** [reply]
Ok that sounds good. I'll look into it after a bit. Thanks for your help!

**Comment by vigoux** [comment]
For the solution, and the api change, as we are directly calling the lib, any change in the lib we'll be usable by us, but not crate changes.

**Reply by maxbrunsfeld** [reply]
Yeah, all of the needed functionality is already exposed by the core library. We'd just need to change the `tree-sitter-highlight` crate in order to be able to test this kind of behavior at the command line with `tree-sitter highlight`, or to incorporate the same functionality into syntax highlighting on GitHub.

**Reply by vigoux** [reply]
Agreed
